src/common/config.cc

```diff
@@ -604,11 +606,42 @@ void Config::UpdateUlMCS(const json& ul_mcs) {
   ul_mod_order_ = static_cast<size_t>(pow(2, ul_mod_order_bits_));
   InitModulationTable(this->ul_mod_table_, ul_mod_order_);
 
+  double ul_code_rate = ul_mcs.value("code_rate", 0.333);
+  if (ul_code_rate >= 1 || ul_code_rate <= 0)
+    std::cerr << "Invalid UL code rate.. Must be in range (0, 1)." << std::endl;
   uint16_t base_graph = ul_mcs.value("base_graph", 1);
-  uint16_t zc = ul_mcs.value("Zc", 72);
+
+  //uint16_t zc = ul_mcs.value("Zc", 72);
+  //size_t num_rows = ul_mcs.value("nRows", (base_graph == 1) ? 46 : 42);
+  std::vector<size_t> zc_vec = {
+      2,   4,   8,   16, 32, 64,  128, 256, 3,   6,   12,  24, 48,
+      96,  192, 384, 5,  10, 20,  40,  80,  160, 320, 7,   14, 28,
+      56,  112, 224, 9,  18, 36,  72,  144, 288, 11,  22,  44, 88,
+      176, 352, 13,  26, 52, 104, 208, 15,  30,  60,  120, 240};
+  std::sort(zc_vec.begin(), zc_vec.end());
+  size_t zc = SIZE_MAX;
+  size_t i = 0;
+  for (; i < zc_vec.size() - 1; i++) {
+    if (zc_vec.at(i) > 256) {
+      // According to cyclic_shift.cc cyclic shifter for zc
+      // larger than 256 has not been implemented, so we skip them here.
+      break;
+    }
+    if ((zc_vec.at(i) * LdpcNumInputCols(base_graph) * kCbPerSymbol <
+         ofdm_data_num_ * ul_code_rate * ul_mod_order_bits_) &&
+        (zc_vec.at(i + 1) * LdpcNumInputCols(base_graph) * kCbPerSymbol >
+         ofdm_data_num_ * ul_code_rate * ul_mod_order_bits_)) {
+      zc = zc_vec.at(i);
+      break;
+    }
+  }
+  if (zc == SIZE_MAX) zc = zc_vec.back();
+
+  size_t num_rows = int(LdpcNumInputCols(base_graph) / ul_code_rate) -
```

Comment:
int - int = size_t .   Probably a better way to cast this.  Need to Assert on a negative result